[Step2] 연관 관계 매핑 리뷰 요청드립니다.

sojeong-park opened the pull request:
안녕하세요 리뷰어님!
리뷰요청드립니다.
감사합니다 : )

sojeong-park committed:
docs: 연관관계 매핑 기능 추가

sojeong-park committed:
test: 연관관계 매핑 테스트

sojeong-park committed:
feat: 연관관계 매핑 추가

tearight commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -36,65 +42,51 @@ public Answer(Long id, User writer, Question question, String contents) {
     }
 
-    public void setDeleted(boolean deleted) {
+    public void updateDeleted(boolean deleted) {
         this.deleted = deleted;
     }
```
deleted 여부를 파라미터로 받는 것은 결국 setter 와 차이가 없는 것 같은데
deleted 를 되살리는 케이스는 제거해보는 것이 어떨까요 ?
현재 서비스에서도 delete 하는 케이스는 없는 것 같아요.
-> (오타수정)
현재 서비스에서도 delete 를 되살리는 케이스는 없는 것 같아요.

sojeong-park replied:
[updateDeleted 메소드 사용하는 코드](https://github.com/next-step/jwp-qna/pull/56/files#r642542400)
요기서 사용하는 코드라서 유지해야될거 같은데, 
질문에 대한 답변도 함께 삭제 되는 경우니까 Question에서 같이 처리하는게 좋다는 말씀일까요?.?

tearight commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -12,8 +12,14 @@
-    private Long writerId;
-    private Long questionId;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "writer_id", foreignKey = @ForeignKey(name = "fk_answer_writer"))
+    private User writer;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "question_id", foreignKey = @ForeignKey(name = "fk_answer_to_question"))
+    private Question question;
```
ManyToOne 의 default FetchType 은 Eager 로 설정되어 있는데요.
어떠한 점을 고려하여 Lazy 로 설정하신 것인지 궁금합니다.
또한 Lazy 를 사용할 때에는 조심해야 할 점이 있을까요 ?

sojeong-park replied:
1. Lazy 설정 이유

- 즉시 (Eager) 로딩은 JPQL에서 N+1 문제가 발생해 어떤 쿼리가 나갈지 예측이 어렵다고 알고있어요.
- Question객체와 연관된 Answer와 User 객체를 Eager 로딩을 사용해 하나의 조인된 쿼리문으로 조회해오면 편하겠지만, 실무에서는 위와 같은 이유로 Eager로딩 사용을 지양하고 있어서 Lazy로딩으로 설정해보았습니다.

2. Lazy 사용시 조심해야할 점

- Lazy 로딩 설정을 한 객체는 프록시 객체를 이용해 조회를 수행하는데, 이때 영속성 컨택스트가 끝난다음에 프록시 객체를 사용하면 LazyInitializationException 이 발생해 주의가 필요하다고 알고있습니다.
- 주로 트랜잭션이 시작하고 끝날때 영속성 컨텍스트도 시작하고 끝나는데 트랜잭션이 끝나고 프록시 객체를 조회하면 LazyInitializationException 이 발생하니 트랜잭션 내에서 Lazy 로딩으로 설정한 객체를 사용해야 한다는 점이 Lazy 로딩 사용시 주의사항이라 생각됩니다.

tearight commented on src/main/java/qna/domain/Question.java:
```diff
@@ -18,7 +16,11 @@
 
     @Lob
     private String contents;
```
제가 이전 단계에서 놓친 부분 같네요.
contents 또한 nullable 인지 확인해야 할 부분 같네요.
혹시 null 이 가능한 요소라 판단하셔서 nullable 을 기재 안하신 것일까요 ?

tearight replied:
이 부분은 제가 ddl 을 잘못 읽어서 착각이 있었네요. 🙇🏻
답변 안남겨주셔도 될 것 같습니다 !

tearight commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,8 +1,6 @@
-
 import javax.persistence.*;
 
 @Entity
```
next-step 2단계 강의 페이지에서 힌트 부분에 아래와 같은 글이 있는데요.
(https://edu.nextstep.camp/s/ffeVDScX/ls/gULJdWFU)

![image](https://user-images.githubusercontent.com/17865104/120200131-f353c580-c25e-11eb-847c-07d2180a3b7c.png)

현재 코드에서는 Question 의 Answer 를 찾기 위해서는 별도의 repository 를 통해야만 하는걸로 보여요.
Question 이라는 객체의 탐색을 통해 Answer 를 찾을 수 있게 개선이 필요해 보입니다.

tearight replied:
QnA Service 에서도 repo 를 참조하고 있는데 이 부분도 같이 수정을 해주셔야 할 것 같아요.

tearight replied:
다른 객체들간의 관계에서도 객체를 통해 탐색할 수 있는 것이 아직 repo 를 통하도록 되어 있는지 확인해보시면 좋을 것 같아요.

tearight reviewed (changes requested):
안녕하세요.
2단계 PR 에 대해 몇가지 코멘트를 남겼습니다.
질문과 개선 포인트 두 가지를 남겼는데요.
질문에 대해서는 답변만 남겨주시면 될 것 같습니다 :D

sojeong-park commented on src/main/java/qna/service/QnAService.java:
```diff
@@ -47,11 +47,11 @@ public void deleteQuestion(User loginUser, Long questionId) throws CannotDeleteE
         for (Answer answer : answers) {
-            answer.setDeleted(true);
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriterId(), LocalDateTime.now()));
+            answer.updateDeleted(true);
```
Answer 클래스의 updateDeleted() 는 여기서 삭제 상태값을 업데이트 하는데 필요한 메소드 같습니다.
유지해야될 메소드 같아요..!

tearight replied:
파라메터가 false 로 넘어가는 경우가 현재 없는 것 같은데 아예 파라메터가 없도록 하면 어떨까요?

tearight replied:
`question.delete()`
`answer.delete()`

이 방식으로 호출하도록 처리해보면 어떨까 제안드려보았어요.
내부는 `this.deleted = true`
이런식으로 구현하게 될 것 같네요.

단순 delete 를 true, false 로 바꾸는 것이기 때문에 `updateDeleted(boolean deleted)` 이런식으로 하면
setter 가 사라지지 않을까 하고 고민하는 경우가 많은데요. 사실 setter 이름만 바뀌는 것이기 때문에 setter 가 사라지지 않은 상태라고 보시면 될 것 같아요.

그렇다면 반대로 삭제된 글을 되살려야 할 경우에는 `question.restore()`라고 표현할 수 있는데요.
restore 는 복구라는 뜻이죠. updatedDeleted(false) 보다 좀 더 도메인적인 언어로 행위가 표현되는 것이죠.

앞으로도 setter 를 제거하는데에 이런 케이스가 많이 존재할 수 있는데요.
단순 boolean 값 뿐만 아니라 enum 등의 속성을 변경할 때도 응용할 수 있는 개념이기도 해요.
Order.updateState(OrderState.WAIT) 보다 Order.wait() 로 표현할 수 있기도 하죠.
이렇게 할 경우 또 다른 장점은 원하지 않는 상태변경(다른 State 를 지정하는 것을 막음) 을 막을 수 있다는 것이에요.

sojeong-park committed:
docs: Question과 Answer 양방향 매핑 수정

sojeong-park committed:
test: Question 객체를 이용해 Answer 객체 탐색 테스트

sojeong-park committed:
refactor: Question 객체를 이용해 Answer 객체 탐색 테스트

Question 객체를 이용해 Answer 객체 탐색 테스트
- 양방향 매핑 추가
- setter 를 delete 상태를 true로만 지정만 하는 메소드로 변경

tearight reviewed (approved):
안녕하세요, 피드백 반영 잘 해주셨습니다. 👍
그럼 3단계에서 뵙겠습니다.
